vm/src/builtins/pytype.rs

```diff
@@ -282,11 +282,20 @@ impl PyType {
 
     pub fn name(&self) -> BorrowedValue<str> {
         PyRwLockReadGuard::map(self.slots.name.read(), |slot_name| {
-            slot_name.as_ref().unwrap().rsplit('.').next().unwrap()
+            slot_name.as_ref().unwrap().as_str()
```

Comment:
In my thought, this change should be reverted because it makes a break between RustPython and CPython.

There is an example to check it. You can run `binascii.Incomplete.__name__` in two environments.

```python
# CPython
>>> import binascii; binascii.Incomplete.__name__
'Incomplete'

# RustPython
>>>>> import binascii; binascii.Incomplete.__name__
'binascii.Incomplete'
```

You can see the `name` CPython implementation in:

https://github.com/python/cpython/blob/8a310dd5f4242b5d28013c1905c6573477e3b957/Objects/typeobject.c#L489

Reply:
Great point, thanks! Looking at that typeobject.c implementation, I think the existing implementation is implementing one of these branches, and I have (incorrectly) *replaced* that implementation with an implementation of the other branch. Each is correct for different cases, which is why the tests pass with my implementation, but your counterexample is broken by it. I'm on my phone at the moment, so I can't dig into it further but: I agree, there's no point landing this change as-is. I'll look into an implementation of *both* branches.

Reply:
Hmm, I actually think my initial analysis was incorrect here (at least for the specific `binascii` example). In CPython, `Incomplete` is created by calling `PyErr_NewException("binascii.Incomplete", ...)` (https://github.com/python/cpython/blob/main/Modules/binascii.c#L1257). `PyErr_NewException` _splits_ the given string on that `.`: the first half is set as `__module__`, the second half is set as `__name__`.

So CPython's `binascii.Incomplete` object only has `Incomplete` as its name: this would be compatible with my implementation. So I think this is an example of a class of issue we'll have: RustPython exceptions will currently store `module.name` as `__name__`, which isn't how CPython Exceptions are constructed.

I'll dig into this some today.

Reply:
Oh, https://github.com/python/cpython/blob/main/Python/errors.c#L1082-L1087 is the relevant `PyErr_NewException` code.

Reply:
https://github.com/RustPython/RustPython/pull/3396 is what I've come up with to fix up the `binascii.Incomplete` case (and those like it). Please do let me know if you think that's sufficient to unblock landing this, or if there's more you'd like to see!

Reply:
I have researched after I saw the commit message of 07cfd112917d0438bda551875fdd5e2253fbfe65, which I missed in the before review.

First, [the fixup](#3396) looks good. 👍🏻 

But, in my current thought, this line should be reverted and we should add other lines to resolve the problem. 
CPython implementation seems to have two variables related to its name, `PyHeapTypeObject.ht_name` and `PyTypeObject.tp_name`. When we see the `type_name` function implementation:

```c
static PyObject *
type_name(PyTypeObject *type, void *context)
{
    puts("type_name"); // Added by @moreal for debugging
    if (type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
        puts("is heaptype"); // Added by @moreal for debugging
        puts(type->tp_name); // Added by @moreal for debugging
        PyHeapTypeObject* et = (PyHeapTypeObject*)type;

        Py_INCREF(et->ht_name);
        return et->ht_name;
    }
    else {
        puts("is not heaptype"); // Added by @moreal for debugging
        return PyUnicode_FromString(_PyType_Name(type));
    }
}
```

Then, let's execute `import _ssl; _ssl.SSLError.__name__` and you will see the below outputs:

```
type_name
is heaptype
ssl.SSLError
'SSLError'
```

It branches into `type->tp_flags & Py_TPFLAGS_HEAPTYPE` and it returns `ht_name`.

Then, in case the commit message of 07cfd112917d0438bda551875fdd5e2253fbfe65:

```python
>>> class C: ...
... 
>>> C.__name__
type_name
is heaptype
C
'C'
>>> C.__name__ = "A.B"
>>> C.__name__
type_name
is heaptype
A.B
'A.B'
>>> 
```

The CPython implementation updates the `ht_name` field at [Objects/typeobject.c#L541](https://github.com/python/cpython/blob/14a4fce457033412278ca9a056787db424310dc3/Objects/typeobject.c#L541), in `type_set_name` (i.e. `PyType::set_name`), and at [Objects/typeobject.c#L3426](https://github.com/python/cpython/blob/b494f5935c92951e75597bfe1c8b1f3112fec270/Objects/typeobject.c#L3426), in `PyType_FromModuleAndSpec` (like `ssl.SSLError`).

Maybe, we should add new property like `ht_name` and separate the role.

Reply:
Thanks for this comment, it's really helpful! I should have some time over the weekend to get my head back into this.

Reply:
I found this is resolved by d531e90319